test/util/malicious/app.go

```diff
@@ -38,24 +62,30 @@ func New(
 	goodApp := app.New(logger, db, traceStore, loadLatest, skipUpgradeHeights, homePath, invCheckPeriod, encodingConfig, appOpts, baseAppOptions...)
 	badApp := &App{App: goodApp}
 
-	// default to using the good app's handlers
-	badApp.SetPrepareProposalHandler(goodApp.PrepareProposal)
-
-	// override the handler if it is set in the app options
-	if prepareHander := appOpts.Get(PrepareProposalHandlerKey); prepareHander != nil {
-		badApp.SetPrepareProposalHandler(prepareHander.(func(req abci.RequestPrepareProposal) abci.ResponsePrepareProposal))
+	// set the malicious prepare proposal handler if it is set in the app options
+	if malHanderName := appOpts.Get(BehaviorConfigKey); malHanderName != nil {
+		badApp.SetMaliciousBehavor(malHanderName.(BehaviorConfig))
 	}
 
 	return badApp
 }
 
-func (app *App) PrepareProposal(req abci.RequestPrepareProposal) abci.ResponsePrepareProposal {
-	return app.preparePropsoalHandler(req)
+func (app *App) SetMaliciousBehavor(mcfg BehaviorConfig) {
+	// check if the handler is known
+	if _, ok := app.PrepareProposalHandlerMap()[mcfg.HandlerName]; !ok {
+		panic("unknown malicious prepare proposal handler")
+	}
+	app.malPreparePropsoalHandler = app.PrepareProposalHandlerMap()[mcfg.HandlerName]
```

Comment:
[Non-blocking]
Would this support multiple malicious handlers to be activated at different times?

Reply:
that is a good question. Originally this was not designed for, however if the chain was halted (easy with one val), then we could add extra code that could call this or a similar function multiple times depending on height.